### PR TITLE
ACLdex, first iteration of abstraction

### PR DESCRIPTION
**Pull request by shaneding**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,10 +2,12 @@
 *.pyc
 __pycache__
 .pytest_cache/
+api/.DS_Store
+.DS_Store
 
 # Anserini index
 anserini/
-api/index/
+# api/index/
```
Comment (edwinzhng) [inline .gitignore:10]: keep this?

File: lucene_schema.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (edwinzhng) [inline lucene_schema.json:1]: acl_schema.json?

can you make a similar one for CORD-19 with all the fields we have now

File: client/src/shared/Constants.ts [hunk 3]
```diff
@@ -14,7 +14,8 @@ export const CONTENT_WIDTH = 1100;
 
 /* API */
 export const API_BASE =
-  process.env.NODE_ENV === 'development' ? 'http://23.251.153.91:8000/api' : '/api';
+  process.env.NODE_ENV === 'development' ? 'http://localhost:8000/api' : '/api';
+  // process.env.NODE_ENV === 'development' ? 'http://23.251.153.91:8000/api' : '/api';
```
Comment (edwinzhng) [inline client/src/shared/Constants.ts:18]: just delete this line

File: api/app/models.py [hunk 4]
```diff
@@ -1,21 +1,23 @@
...
-    source: List[str] = []
-    url: str
+def gettype(name):
+    t = getattr(__builtins__, name)
+    if isinstance(t, type):
+        return t
+    raise ValueError(name)
 
 
+lucene_schema = json.load(open("../lucene_schema.json"))
```
Comment (edwinzhng) [inline api/app/models.py:16]: put the file name into an env variable, and move all the JSON files into a `schema` folder? then you'll have `acl.json`, `cord19.json` which you could specify in the `.env` file

File: api/app/settings.py [hunk 5]
```diff
@@ -9,14 +9,9 @@ class Settings(BaseSettings):
     development: bool = True
     log_path: str = 'logs/'
 
-    # Related searcher settings
-    related_bin_path: str = "index/cord19-hnsw-index/cord19-hnsw.bin"
```
Comment (edwinzhng) [inline api/app/settings.py:13]: Let's keep these, I'll fix them later

File: lucene_schema.json (hunk 2 above)
Comment (edwinzhng) [inline lucene_schema.json:1]: delete this file?

File: client/src/App.tsx [hunk 6]
```diff
@@ -5,7 +5,7 @@ import styled, { ThemeProvider } from 'styled-components';
 import Navbar from './components/navigation/Navbar';
 import Footer from './components/navigation/Footer';
 import HomePage from './components/pages/HomePage/HomePage';
-import RelatedPage from './components/pages/RelatedPage/RelatedPage';
+// import RelatedPage from './components/pages/RelatedPage/RelatedPage';
```
Comment (edwinzhng) [inline client/src/App.tsx:8]: I think you can leave this in? Shouldn't be any harm as long as we dont show the links anywhere

File: client/src/components/common/AclBaseArticleResult.tsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  let authorString = '';
+  if (article.authors.length > 0) {
+    article.authors.forEach((author, idx) => {
+      if (author !== '') {
+        authorString += idx === article.authors.length - 1 ? `${author}.` : `${author}, `;
+      }
+    });
+  }
+
+  const transformPreprintSource = (source: string) => {
```
Comment (edwinzhng) [inline client/src/components/common/AclBaseArticleResult.tsx:29]: Won't need this for ACL (lines 29-43), all the papers will be published at venues, not as preprints

File: client/src/components/common/BaseArticleResult.tsx [hunk 8]
```diff
@@ -1,11 +1,12 @@
+//@ts-nocheck
```
Comment (edwinzhng) [inline client/src/components/common/BaseArticleResult.tsx:1]: why is this here

Comment (shaneding) [reply client/src/components/common/BaseArticleResult.tsx:1]: typescript throws an error because BaseAclArticle has no journal field 
```
{article.journal && <Journal>{article.journal}</Journal>}
```

Comment (shaneding) [reply client/src/components/common/BaseArticleResult.tsx:1]: Added so it ignores error in `BaseArticleResult`

Comment (edwinzhng) [reply client/src/components/common/BaseArticleResult.tsx:1]: Don't keep the subtitle rendering in this base component then, move it out into an abstract `renderSubtitle()` function that is implemented specifically for each dataset

File: client/src/components/pages/HomePage/AclSearchResult.tsx [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+      searchWords={queryTokens}
+      textToHighlight={text}
+      highlightTag={Match}
+      highlightClassName="match"
+    />
+  );
+};
+
+// adjust highlights based on difference
+const adjustHighlights = (
```
Comment (edwinzhng) [inline client/src/components/pages/HomePage/AclSearchResult.tsx:87]: can you move all these duplicated highlight functions into a `utils/highlight.tsx` file and import them?

File: client/src/components/pages/HomePage/SearchResult.tsx [hunk 10]
```diff
@@ -100,9 +101,7 @@ const SearchResult = ({ article, position, queryId, queryTokens }: SearchResultP
   const [collapsed, setCollapsed] = useState<boolean>(true);
 
   // Separate abstract from other paragraphs if it was highlighted
-  const originalAbstract = article.highlighted_abstract
-    ? article.paragraphs[0]
-    : article.abstract || '';
+  const originalAbstract = article.abstract_html || '';
```
Comment (edwinzhng) [inline client/src/components/pages/HomePage/SearchResult.tsx:104]: rename this class to Cord19SearchResult and revert all the changes?

Comment (edwinzhng) [reply client/src/components/pages/HomePage/SearchResult.tsx:104]: i dont think you need to touch this since the user should just choose which type of search result to render

File: client/src/shared/Models.ts [hunk 11]
```diff
@@ -1,23 +1,33 @@
...
+export interface SearchArticle extends BaseAclArticle {
   highlights: Array<Array<[number, number]>>;
   highlighted_abstract: boolean;
   paragraphs: Array<string>;
   score: number;
   has_related_articles: boolean;
 }
 
-export interface RelatedArticle extends BaseArticle {
+export interface RelatedArticle extends BaseAclArticle {
```
Comment (edwinzhng) [inline client/src/shared/Models.ts:30]: Change this to `Cord19RelatedArticle extends BaseCord19Article` since there is not related articles for ACL

File: client/src/components/common/BaseArticleResult.tsx (hunk 8 above)
Comment (edwinzhng) [reply client/src/components/common/BaseArticleResult.tsx:1]: Actually just thinking about it more, we should have two entirely separate `Cord19ArticleResult`, `AclArticleResult` classes that take the same props interface

Comment (edwinzhng) [reply client/src/components/common/BaseArticleResult.tsx:1]: Don't bother extending them off one base class, so just define like `ArticleResultProps` in a Common.tsx file or something and import it into both

File: schema/acl_schema.json [hunk 12]
```diff
@@ -0,0 +1,43 @@
+{
```
Comment (edwinzhng) [inline schema/acl_schema.json:1]: do you mind moving this file under /api since it's only used on the backend

File: client/src/shared/acl_filter.json [hunk 13]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (edwinzhng) [inline client/src/shared/acl_filter.json:1]: Can you make a cord19_filter.json as well for completeness

File: client/src/shared/Constants.ts [hunk 14]
```diff
@@ -1,4 +1,5 @@
 import { OptionTypeBase } from 'react-select';
+import json_schema from './acl_filter.json'
```
Comment (edwinzhng) [inline client/src/shared/Constants.ts:2]: conventionally, we don't use snake_case in javascript, maybe just rename this to `FilterJsonSchema` and `FilterSchema` below as well?
